[MRG] Fast, low memory, single linkage implementation

lmcinnes opened the pull request:
#### Reference Issues/PRs
Fixes #11513


#### What does this implement/fix? Explain your changes.
This PR provides a small cython routine for computing the core mst computation in single linkage clustering in a way that computes distances only as needed, releasing them when they have been used. This allows for single linkage clustering on vector data to run successfully without exploding memory. As a side benefit this provides a speed improvement over the single linkage clustering implementation available in scipy.

#### Any other comments?
I originally sought to add this upstream to scipy. However scipy's distance computations are not exposed to cython, so on the fly distance computations are not feasible. In constrast Jake Vanderplas implemented most distance functions in the neighbors package in Cython, making fast computation available for all of these distance metrics easily implementable in scikit-learn.

lmcinnes committed:
First cut at basic single linkage internals

lmcinnes committed:
Refer to correct dist_metrics package

lmcinnes committed:
Add csgraph sparse implementation for single linkage

lmcinnes committed:
Add fast labelling/conversion from MST to single linkage tree; remove uneeded single_linkage.pyx file.

lmcinnes committed:
Ensure existing tests cover single linkage

lmcinnes committed:
Name cingle linkage labelling correctly.

lmcinnes committed:
Iterating toward correct solution. Still have to get n_clusters, compute_full_tree=False working

lmcinnes committed:
Get n_components correct.

lmcinnes committed:
Update docstrings.

lmcinnes committed:
Fix the parents array when we don't get the "full tree"

lmcinnes committed:
Add single linkage to agglomerative clustering example.

lmcinnes committed:
Add single linkage to digits agglomerative clustering example.

lmcinnes committed:
Update documentation to reflect the addition of single linkage.

lmcinnes committed:
Update documentation to reflect the addition of single linkage.

lmcinnes committed:
Pep8 fix for class declaration in cython

lmcinnes committed:
Fix heading in clustering docs

lmcinnes committed:
Update the digits clustering text to reflect the new reality.

lmcinnes committed:
Provide a more complete comparison of the different linkage methods, highlighting the relative strengths and weaknesses.

lmcinnes committed:
We don't need connectivity here, and we can ignore issues with warnings for spectral clustering.

lmcinnes committed:
Add an explicit test that single linkage successfully works on examples it should perform well on.

lmcinnes committed:
Update docs with a more complete comparison on linkage methods (scale to be determined?)

lmcinnes committed:
List formatting in example linkage comparison.

lmcinnes committed:
Flake8 fixes.

lmcinnes committed:
Flake8 fixes.

lmcinnes committed:
More Flake8 fixes.

lmcinnes committed:
Fix agglomerative plot example with correct subplot spec

lmcinnes committed:
Explicitly test linkages (including single) produce results identical to scipy.cluster.hierarchical

lmcinnes committed:
Fix comment on why we sort (consistency)

lmcinnes committed:
Make dense single linkage faster

lmcinnes committed:
Merge branch 'master' into fast_single_linkage

# Conflicts:
#	doc/modules/clustering.rst
#	examples/cluster/plot_linkage_comparison.py
#	sklearn/cluster/_hierarchical.pyx
#	sklearn/cluster/hierarchical.py
#	sklearn/cluster/tests/test_hierarchical.py

lmcinnes committed:
Add docstring to new mst-linkage-core computations.

lmcinnes committed:
Add a test that new single linkage code matches scipy

lmcinnes committed:
Ensure we only attemtp this for metrics Jake implemented.

amueller commented:
Usually it should be faster to pre-compute, right? Should we decide based on data size? Or never precompute? We had a similar issue with dbscan where basically we said that maybe we want to allow chunked computation of the k-neighbors graph and then pass the sparse matrix. I guess that would be slower than this? I'm not 100% sure, though....

amueller commented:
Oh never mind, this algorithm is smarter than i thought. Benchmark would still be appreciated ;)

amueller commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -448,4 +451,119 @@ def single_linkage_label(L):
+    tree for computation of single linkage clustering. This
+    represents the MST-LINKAGE-CORE algorithm from
+    *Modern hierarchical, agglomerative clustering algorithms*
+    by Daniel Mullner (https://arxiv.org/abs/1109.2378).
```
maybe add it's figure 6?

lmcinnes replied:
Done, thanks.

amueller replied:
I'm just very lazy ;)

lmcinnes committed:
Per amueller; it's a long paper, ref the figure.

lmcinnes committed:
Clean up a few things.

lmcinnes committed:
Too many blank lines for flake8

lmcinnes committed:
Bad scipy slink input

lmcinnes committed:
Flake8 fixes

lmcinnes committed:
Clean up cython a little; fix typo/carryover

lmcinnes committed:
Convert memoryview to numpy array on return

lmcinnes commented:
The plots above were for 10 dimensional data; here's the same for 100 dimensional data. 

![image](https://user-images.githubusercontent.com/11962885/42735699-8940cc60-881e-11e8-8f17-7a7e96f3352d.png)

![image](https://user-images.githubusercontent.com/11962885/42735708-905ec2e0-881e-11e8-94ed-1e2e2936d700.png)

The speed improvements decline, but the low memory aspect is still a distinct improvement. It is likely that for small but very high dimensional data one may want to compute the full distance matrix and work from there. I'll try something 10000 dimensional and get back to you.

amueller commented:
Cool. Just making sure we have all bases covered. Thanks for the investigation!

lmcinnes commented:
Benchmarks for 2000 dimensional:

![image](https://user-images.githubusercontent.com/11962885/42977611-27cee09a-8b96-11e8-915e-b3f420fa95b9.png)

and log-log for completeness:

![image](https://user-images.githubusercontent.com/11962885/42977617-3432c3ce-8b96-11e8-80ca-39db962d8068.png)

Technically the new version is marginally slower here, but it's a close call, and personally I'm dubious of the value of clustering this high dimensional a dataset.

Also I don't understand the LGTM errors. If anyone can help me figure out what's going wrong there I would appreciate it.

jnothman reviewed (commented):
This looks fairly complete if we want it.

The memory improvements here sound attractive (although no one has complained, tml). As suggested by @amueller, a precomputed sparse neighborhood graph might also be usable in single linkage, and is the memory saver we've chosen elsewhere.

Without looking into details I assume this changes the linkage but not the clustering in cases of ties?

lmcinnes commented:
@jnothman : I certainly agree that a sparse neighbor graph is a general solution to memory issues, and is indeed what is often used elsewhere. There are some potential issues in that it doesn't guarantee identical results to the full computation. Right now single linkage supports the sparse case, but pushes the responsibility for recognising potential issues to the user (who has to do the sparsification themselves). Perhaps when the new nearest neighbor transformer gets merged it could get used here? In the meantime I think it would be useful, from a principle of least surprise view, to let the user throw the data in as is and still get a result if at all possible.

Finally, yes, the linkage may differ in the case of ties, but the clustering results will be the same.

lmcinnes committed:
Merge branch 'master' into fast_single_linkage

amueller commented:
I think this is useful and maybe @NicolasHug has time to take a look at it? I probably won't :-/

lmcinnes committed:
Just convert to the correct dtype

NicolasHug commented on sklearn/cluster/hierarchical.py:
```diff
@@ -464,7 +466,25 @@ def linkage_tree(X, connectivity=None, n_clusters=None, linkage='complete',
+        if (linkage == 'single'
+                and affinity != 'precomputed'
+                and not callable(affinity)
+                and affinity in METRIC_MAPPING):
```
It seems that `affinity in METRIC_MAPPING` actually covers the 2 previous conditions

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+        algorithm MST-LINKAGE-CORE for more details.
+    """
+
+    cdef DTYPE_t[:] current_distances
```
nit: we can use a `cdef:` block

and also just initialize variables here during definition instead of doing it later

lmcinnes replied:
Fixing now ...

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+    Parameters
+    ----------
+    raw_data: array of shape (n_samples, n_features)
+        The array of feature data to be clustered.
```
```suggestion
        The array of feature data to be clustered. Must be C-aligned
```

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] mst_linkage_core(
+        np.ndarray[DTYPE_t, ndim=2, mode='c'] raw_data,
```
Should be enough (you're still allowed to pass it actual numpy arrays)

```suggestion
        DTYPE_t [:, ::1] raw_data,
```

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+# Implements MST-LINKAGE-CORE from https://arxiv.org/abs/1109.2378
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] mst_linkage_core(
```
This should just be a `def` since it's never called from cython

lmcinnes replied:
Yes; done.

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+
+    raw_data_view = (<DTYPE_t[:raw_data.shape[0], :raw_data.shape[1]:1]> (
+        <DTYPE_t *> raw_data.data))
+    raw_data_ptr = (<DTYPE_t *> &raw_data_view[0, 0])
```
I think we can avoid this and just declare raw_data as a view from the function signature (see comment above). You'll still have access to `raw_data.shape` and `&raw_data`

NicolasHug commented on sklearn/cluster/hierarchical.py:
```diff
@@ -464,7 +466,25 @@ def linkage_tree(X, connectivity=None, n_clusters=None, linkage='complete',
+
+            mst = _hierarchical.mst_linkage_core(X, dist_metric)
+            # Sort edges of the min_spanning_tree by weight
+            mst = mst[np.argsort(mst.T[2]), :]
```
according to the paper, should this be `kind='stable'`? does it matter?

lmcinnes replied:
I believe it doesn't actually matter since it will come out in the wash with the next step, but it is easy enough to add, and then we can agree with the reference.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -305,6 +305,31 @@ def test_scikit_vs_scipy():
 
 
+def test_vector_scikit_single_vs_scipy_single():
+    n, p, k = 10, 5, 3
```
```suggestion
    n_samples, n_features, n_clusters = 10, 5, 3
```

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+                    new_node = j
+
+        result[i - 1, 0] = <double> current_node
+        result[i - 1, 1] = <double> new_node
```
nit: not sure we need to explicit cast here

lmcinnes replied:
I honestly can't recall; I'll remove it and see what happens.

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+    result = np.zeros((dim - 1, 3))
+    in_tree = np.zeros(dim, dtype=np.int8)
+    current_node = 0
+    current_distances = np.infty * np.ones(dim)
```
`np.full` is preferred for this

Also we typically use ` from numpy.math cimport INFINITY`  for representing infinity.

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+
+        in_tree[current_node] = 1
+
+        new_distance = DBL_MAX
```
use INFINITY as well?

lmcinnes replied:
I assume I had a reason for this, but I'm not sure now. Let's see if INFINITY makes it through the tests; definitely a better option.

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+                    new_node = j
+                continue
+
+            if left_value < right_value:
```
remove continue and use `elif` below?

lmcinnes replied:
Done.

NicolasHug commented on sklearn/cluster/_hierarchical.pyx:
```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+            else:
+                if right_value < new_distance:
+                    new_distance = right_value
+                    new_node = j
```
as far as I understand the logic here, you're doing all this to keep track of new_node = argmin_j(current_distances) right?

I might be missing something but there is probably a way to simplify this a bit?

lmcinnes replied:
It's been a while since I wrote this, but my recollection was that there were simpler approaches, but this was (non-trivially) faster (for, overall, not too much extra complexity). I do know I spent quite a while fiddling with exactly how this worked before arriving at this as the best option.

NicolasHug reviewed (commented):
thanks for the PR @lmcinnes! 

Made a first pass of comments (mostly on the Cython code). Looks pretty good in general, though I'm not familiar with this yet so I'll need more time to read the paper in more details

lmcinnes committed:
Update sklearn/cluster/_hierarchical.pyx

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes committed:
Update sklearn/cluster/_hierarchical.pyx

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes committed:
Update sklearn/cluster/_hierarchical.pyx

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes committed:
Update sklearn/cluster/tests/test_hierarchical.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes commented on sklearn/cluster/hierarchical.py:
```diff
@@ -464,7 +466,25 @@ def linkage_tree(X, connectivity=None, n_clusters=None, linkage='complete',
+            dist_metric = DistanceMetric.get_metric(affinity)
+
+            # The Cython routines used require contiguous arrays
+            X = np.array(X, dtype=np.double, order='C')
```
Makes sense -- fixing now.

lmcinnes committed:
Fixes as per @NicolasHug suggestions.

lmcinnes commented:
@NicolasHug Thanks for taking the time to look at this, and thanks for the feedback. I wrote a lot of this quite some time ago, and my memory is a little hazy as to how or why various things are the way they are. Your suggestions fix up a lot of my poor code style, so I have fixed that where I can. Hopefully I can get most of this resolved soon.

lmcinnes committed:
Merge upstream/master in

lmcinnes committed:
Update renaming of params in test_hierarchical

lmcinnes committed:
Relative import?

lmcinnes committed:
Ah, it got renamed in master...

lmcinnes committed:
A bad merge on my part.

lmcinnes committed:
In principle this is in sklearn.neighbors now...

lmcinnes committed:
No; not that way...

NicolasHug commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
+        ITYPE_t new_node
+        ITYPE_t i
+        ITYPE_t j
+        ITYPE_t dim = raw_data.shape[0]
```
sorry about the painful merged (as you saw we renamed most files and changed imports).  The tests pass on my machine now, you just need to declare dim before it's used

lmcinnes replied:
Ah, excellent; thanks. It's hard when everything starts breaking.

lmcinnes committed:
Declare dim before use.

lmcinnes commented:
Seems like a chunk of the issue is argsort doesn't have ``kind=stable`` prior to version 1.15.0. I suggest we remove the ``kind=stable``.  While this technically changes the result under equality, the next step is labelling with union-find which will simply declare all the merges to happen essentially at once; the resulting hierarchy will be isomorphic regardless of how ties are handled. I am happy to hear alternative options of course.

jnothman commented:
kind='mergesort' is stable

NicolasHug commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
+        algorithm MST-LINKAGE-CORE for more details.
+    """
+    cdef:
+        ITYPE_t dim = raw_data.shape[0]
```
Nit: call this n_samples ?

lmcinnes replied:
Good point.

NicolasHug commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
+
+        DTYPE_t[:] current_distances = np.full(dim, INFINITY)
+
+    for i in range(1, dim):
```
```suggestion
    for i in range(dim - 1):
```

and use `i` instead of `i - 1` below

lmcinnes replied:
I think this is cruft left over from prior iterations of the code. Fixed.

NicolasHug commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
+            else:
+                if right_value < new_distance:
+                    new_distance = right_value
+                    new_node = j
```
I don't think you need the `else` here?

But in any case, it seems to me that that whole section could be replaced by:

```py
            if left_value < right_value:
                current_distances[j] = left_value

            if current_distances[j] < new_distance:
                new_distance = current_distances[j]
                new_node = j
```

This passes the tests

lmcinnes replied:
I had to double check this, but you are quite correct. Again, I think iterating to a solution left things in a less optimal state. Fixes.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -314,6 +314,31 @@ def test_scikit_vs_scipy():
         _hc_cut(n_leaves + 1, children, n_leaves)
 
 
+def test_vector_scikit_single_vs_scipy_single():
```
Small comment like

```py
# Make sure our custom mst_linkage_core gives the same results as scipy's builtin
```

lmcinnes replied:
Done.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -314,6 +314,31 @@ def test_scikit_vs_scipy():
+        X -= X.mean(axis=1)[:, np.newaxis]
+
+        out = hierarchy.linkage(X, method='single')
+        children_ = out[:, :2].astype(np.int)
```
```suggestion
        children_scipy = out[:, :2].astype(np.int)
```

NicolasHug reviewed (commented):
Second pass, looks good!

Regarding ties, the paper says

> This should be taken as a warning that ties demand more from an algorithm and must be explicitly taken into account when we prove the correctness of the MST-linkage algorithm below.

I agree with you that the result should be isomorphic, but I guess using mergesort can't hurt, since it would also ensure consistent results across different calls

lmcinnes committed:
Update sklearn/cluster/tests/test_hierarchical.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes commented:
I'm travelling soon, so I may not be able to get to this for a few days; we'll see how things go.

lmcinnes committed:
Remaining fixes per Nicolas Hug.

lmcinnes committed:
Merge remote-tracking branch 'origin/fast_single_linkage' into fast_single_linkage

# Conflicts:
#	sklearn/cluster/tests/test_hierarchical.py

lmcinnes commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
         raise ValueError("Input MST array must be sorted by weight")
 
     return _single_linkage_label(L)
+  
```
Fixed.

lmcinnes commented on sklearn/cluster/_hierarchical_fast.pyx:
```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
+            left_value = dist_metric.dist(&raw_data_ptr[num_features *
+                                                        current_node],
+                                          &raw_data_ptr[num_features * j],
+                                          num_features)
```
Fixed.

lmcinnes commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -314,6 +314,31 @@ def test_scikit_vs_scipy():
+    n_samples, n_features, n_clusters = 10, 5, 3
+    rng = np.random.RandomState(0)
+
+    for i in range(5):
```
I think I got this fixed -- not used to pytest parameterize, so hopefully I got it right.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -314,6 +314,31 @@ def test_scikit_vs_scipy():
 
+# Make sure our custom mst_linkage_core gives the same results as scipy's builtin
+@pytest.mark.parametrize('seed', range(5))
+def test_vector_scikit_single_vs_scipy_single():
```
you'll need to define the seed as an argument now:

```suggestion
def test_vector_scikit_single_vs_scipy_single(seed):
```

Also there are pep8 issues which will prevent most of the CIs from running, you can check locally by running `make flake8-diff`

lmcinnes committed:
Update sklearn/cluster/tests/test_hierarchical.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes committed:
Fix flake8 issues.

lmcinnes committed:
Merge remote-tracking branch 'upstream/master' into fast_single_linkage

lmcinnes committed:
Switch from stable to mergesort per jnotham

lmcinnes commented:
Remaining errors don't seem to be related to anything I did -- not sure what's going on; I'll try to deal with it when I get done with travels.

NicolasHug commented on sklearn/cluster/_hierarchical.py:
```diff
@@ -18,6 +18,8 @@
 from ..utils import check_array
 from ..utils.validation import check_memory
+from ..neighbors import DistanceMetric
+from ..neighbors.dist_metrics import METRIC_MAPPING
```
This should make CI happy ;)

```suggestion
from ..neighbors._dist_metrics import METRIC_MAPPING
```

lmcinnes committed:
Update sklearn/cluster/_hierarchical.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

lmcinnes commented:
Ah, I see. I probably messed that up in the merge. Sorry.

NicolasHug reviewed (approved):
Last comments but LGTM! 

I can reproduce the benchmarks:
- new implem is marginally slower than scipy's for very small sample sizes, but still very fast anyway so it's fine
- new implem is much faster than scipy's otherwise, and can handle data size where scipy would just run out of memory

Thanks @lmcinnes !!

I'll try to get a second reviewer for this

lmcinnes commented on sklearn/cluster/_hierarchical.py:
```diff
@@ -107,10 +109,10 @@ def _single_linkage_tree(connectivity, n_samples, n_nodes, n_clusters,
 
     # Sort edges of the min_spanning_tree by weight
-    mst_array = mst_array[np.argsort(mst_array.T[2]), :]
+    mst_array = mst_array[np.argsort(mst_array.T[2], kind='mergesort'), :]
```
No, I don't think so -- I was just being extra cautious.

lmcinnes commented on sklearn/cluster/_hierarchical.py:
```diff
@@ -107,10 +109,10 @@ def _single_linkage_tree(connectivity, n_samples, n_nodes, n_clusters,
 
     # Convert edge list into standard hierarchical clustering format
-    single_linkage_tree = _hierarchical._single_linkage_label(mst_array)
+    single_linkage_tree = _hierarchical.single_linkage_label(mst_array)
```
Agreed; I often find more warnings/errors are better, but you are quite correct that the checks are all effectively validated beforehand.

lmcinnes committed:
Skip checks that are already validated.

GaelVaroquaux reviewed (commented):
I had a tiny nitpick, but aside from this, I am +1 for merge.

That said, it would be very good (though not necessary) to have a speed benchmark. It could be added to the bencmarks folder to be able to reproduce later.

lmcinnes committed:
Update docstring per Gael's suggestion

lmcinnes commented:
I'll look into the benchmarks -- presumably something along the lines of the ``bench_plot_fastkmeans.py`` example for hierarchical clustering, potentially comparing different linkage methods?

GaelVaroquaux commented:
> presumably something along the lines of the bench_plot_fastkmeans.py example for hierarchical clustering,

Yes!

> potentially comparing different linkage methods?

You can do this if you want. However, for this PR, what I really need is
something to run to see a meaningful difference between master and this
branch for single linkage. The rest is additional benefits that are not
needed to merge the PR.

Thanks!!!

lmcinnes committed:
Add a benchmark script for agglomerative clustering

lmcinnes committed:
Fix some flake8 issues

GaelVaroquaux commented:
Could you also report the results of the benchmark in the PR, to facilitate review, and for future reference.

Thank you very much!!

lmcinnes committed:
No flake8 on the one line

lmcinnes commented:
Not sure what's up with the one line; the ``bench_plot_fastkmeans.py`` has the same line and the linter doesn't seem to complain in that case.

jnothman commented:
> the bench_plot_fastkmeans.py has the same line and the linter doesn't seem to complain in that case.

We only lint the diff...

lmcinnes committed:
Update parameters and output for benchmark hierarchical

lmcinnes commented:
Benchmarking is done. The results I got are as follows:
Base sklearn: https://gist.github.com/lmcinnes/6469b8d17c14330134f7564397c412b8
This patch: https://gist.github.com/lmcinnes/11cdc47c3403e4ca3422eeeb89dea8c7

![Base plot output](https://user-images.githubusercontent.com/11962885/69364820-0bf22100-0c61-11ea-9412-9188d1186895.png)
![This patch plot output](https://user-images.githubusercontent.com/11962885/69364821-0bf22100-0c61-11ea-93de-8dfe13c176f4.png)

GaelVaroquaux commented:
Thanks heaps!

I have difficulties reading plot in 3D (somewhat ironical, given that I contributed a lot to Mayavi). Would you mind doing those plots in 2D with multiple lines giving the 3rd direction.

lmcinnes commented:
Sure -- I was following the kmeans example, which did 3d plots. I'll try to come up with something reasonable with colour schemes.

lmcinnes committed:
Switch to 2D plotting for hierarchical benchmark

lmcinnes committed:
Wrong colormap name

lmcinnes committed:
Formatting fpr bench hierarchical

lmcinnes commented:
I modified the benchmarking script for 2D plots. This involved a little finessing of the parameters, so I updated the outputs of the scripts:

Base sklearn: https://gist.github.com/lmcinnes/6469b8d17c14330134f7564397c412b8
This patch: https://gist.github.com/lmcinnes/11cdc47c3403e4ca3422eeeb89dea8c7

Plots are as follows:

Base sklearn:

![Base sklearn benchmark](https://user-images.githubusercontent.com/11962885/69390999-614d2300-0c9f-11ea-8a19-e97ef2775f77.png)

This patch:

![This patch benchmark](https://user-images.githubusercontent.com/11962885/69391017-745ff300-0c9f-11ea-8c66-ac76e5f9b511.png)

The final results are essentially as described above.

GaelVaroquaux commented:
Ball-park estimate, speed-up by a factor ~ 1.5x in this branch, right?

GaelVaroquaux commented:
Alright, I had a last read at the code. It's nice and compact. Excellent job. Merging! Thank you!

GaelVaroquaux commented:
Wait! I spoke too early. Can you please update the changelog (in doc/whats_new).

Thanks!

NicolasHug commented:
Needs a what's new entry before merging ;)

lmcinnes committed:
Add an item to WhatsNew

GaelVaroquaux reviewed (approved):
LGTM. Merging! Thank you!

amueller commented:
Awesome!

ogrisel commented:
I put the 0.22 label on this PR because it is currently included in #15715 but @jnothman stated that this should not be part of the 0.22 because it was merged after the feature freeze and 0.22.X branch (and it is not a bugfix). I am +0 to move it to 0.23.

Feel free to retag to 0.23 if everyone agrees.

jnothman commented:
I don't feel strongly about it, really. I just feel like there's an ideal release process where RC means no more features, and you need to draw that line somewhere.

GaelVaroquaux commented:
> I don't feel strongly about it, really. I just feel like there's an ideal release process where RC means no more features, and you need to draw that line somewhere.

I agree. Sorry for merging this. I thought that there was a release branch. Sorry, I should have checked.

jnothman commented:
There was a release branch, but the change log entry here is in 0.22, not
0.23. a mere oversight

lmcinnes commented:
Sorry -- my mistake; I got the wrong what's new file apparently. Is there anything I need to do to correct this?

jnothman commented:
You can submit a pr moving the entry. I don't think it was your mistake in
that we don't usually expect contributors to be aware of the release
cycle's intricacies.
